File: 프로그래머스/2/87946. 피로도/피로도.py
def backtrack(k, dungeons, visited):
    max_cnt = 0
    
    for i, (min_f, use_f) in enumerate(dungeons):
        if not visited[i] and k >= min_f:
            visited[i] = True
            
            max_cnt = max(max_cnt, backtrack(k - use_f, dungeons, visited) + 1)
            
            visited[i] = False
            
    return max_cnt


def solution(k, dungeons):
    visited = [False] * len(dungeons)
    
    return backtrack(k, dungeons, visited)

File: 프로그래머스/2/87946. 피로도/피로도.kt
class Solution {
    fun dfs(k: Int, dungeons: Array<IntArray>, visited: BooleanArray, cnt: Int): Int {
        var maxCnt = cnt
        
        for ((i, dungeon) in dungeons.withIndex()) {
            val minFati = dungeon[0]
            val useFati = dungeon[1]
            
            if (!visited[i] && k >= minFati) {
                visited[i] = true
                
                maxCnt = maxOf(maxCnt, dfs(k - useFati, dungeons, visited, cnt + 1))
                
                visited[i] = false
            }
        }
        
        return maxCnt
    }
    
    fun solution(k: Int, dungeons: Array<IntArray>): Int {
        val visited = BooleanArray(dungeons.size)
        
        return dfs(k, dungeons, visited, 0)
    }
}